lib/pg_eventstore/extensions/options_extension.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+require 'set'
+
+module PgEventstore
+  module Extensions
+    # A very simple extension that implements a DLS for adding attr_accessors with default values,
+    # and assigning their values during object initialization.
+    # Example. Let's say you frequently do something like this:
+    # ```ruby
+    # class SomeClass
+    #   attr_accessor :attr1, :attr2, :attr3, :attr4
+    #
+    #   def initialize(opts = {})
+    #     @attr1 = opts[:attr1] || 'Attr 1 value'
+    #     @attr2 = opts[:attr2] || 'Attr 2 value'
+    #     @attr3 = opts[:attr3] || do_some_calc
+    #     @attr4 = opts[:attr4]
+    #   end
+    #
+    #   def do_some_calc
+    #   end
+    # end
+    #
+    # SomeClass.new(attr1: 'hihi', attr4: 'byebye')
+    # ```
+    #
+    # You can replace the code above using the OptionsExtension:
+    # ```ruby
+    # class SomeClass
+    #   include PgEventstore::Extensions::OptionsExtension
+    #
+    #   option(:attr1) { 'Attr 1 value' }
+    #   option(:attr2) { 'Attr 2 value' }
+    #   option(:attr3) { do_some_calc }
+    #   option(:attr4)
+    # end
+    #
+    # SomeClass.new(attr1: 'hihi', attr4: 'byebye')
+    # ```
+    module OptionsExtension
+      module ClassMethods
+        # @param opt_name [Symbol] option name
+        # @param blk [Proc] provide define value using block. It will be later evaluated in the
+        #   context of your object to determine the default value of the option
+        # @return [Symbol]
+        def option(opt_name, &blk)
+          self.options = (options + Set.new([opt_name])).freeze
+          attr_writer opt_name
+
+          define_method opt_name do
```

Comment:
don't you want to prefix somehow options methods, I'm wondering about potential collisions If someone already has a method that is named as option.

When I have definitions like: 
`def initialize(opts = {})`  I can provide `new(name: 'xyz')` at the same time I can have `def self.name` method in this implementation there will be a collision

Reply:
Could you provide an example of such collisions please? Because I don't see any problem of not using this module in user's specific implementation.

Reply:
As discussed - we can output warnings if reader/writer methods already exist